--- ttutil.c.orig	2010-03-13 00:33:29.000000000 -0500
+++ ttutil.c	2010-03-13 00:34:05.000000000 -0500
@@ -127,7 +127,7 @@ int ttopensockunix(const char *path){
   struct sockaddr_un saun;
   memset(&saun, 0, sizeof(saun));
   saun.sun_family = AF_UNIX;
-  snprintf(saun.sun_path, SOCKPATHBUFSIZ, "%s", path);
+  strncpy(saun.sun_path, path, sizeof(saun.sun_path)-1);
   int fd = socket(PF_UNIX, SOCK_STREAM, 0);
   if(fd == -1) return -1;
   int optint = 1;
@@ -187,7 +187,7 @@ int ttopenservsockunix(const char *path)
   struct sockaddr_un saun;
   memset(&saun, 0, sizeof(saun));
   saun.sun_family = AF_UNIX;
-  snprintf(saun.sun_path, SOCKPATHBUFSIZ, "%s", path);
+  strncpy(saun.sun_path, path, sizeof(saun.sun_path)-1);
   int fd = socket(PF_UNIX, SOCK_STREAM, 0);
   if(fd == -1) return -1;
   if(bind(fd, (struct sockaddr *)&saun, sizeof(saun)) != 0 ||